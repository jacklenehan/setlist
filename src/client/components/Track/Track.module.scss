.root {
  display: flex;
  cursor: move;
  margin-bottom: 0.5em;
  padding: 0.25em;
  justify-content: space-between;
  background: #f0f0f0;

  button {
    width: 2rem;
    padding: 0;
    flex-shrink: 0;
  }

  &:last-child {
    margin-bottom: 0;
  }
}

.info {
  display: flex;
  flex-direction: row;

  > * {
    margin-right: 0.5em;
  }
}

.artistTrack {
  display: block;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  flex-grow: 1;
}

.tempo {
  margin-right: 0.5em;
  width: 3em;
  background: #ddeaff;
  border: none;
  font-size: 14px;

  &:focus {
    outline: none;
  }
}

.dragging {
  opacity: 0.25;
}
