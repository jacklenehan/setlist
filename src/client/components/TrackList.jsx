import React from 'react';
import PropTypes from 'prop-types';

import '../styles/TrackList.css';

const TrackList = ({
  name,
  tracks,
  isTooLong,
  formattedDuration
}) => {
  const className = (isTooLong) ? 'duration too-long' : 'duration';

  return (
    <section className={`TrackList ${name}`}>
      <h3>
        {name}&nbsp;
        <span className={className}>
          ({formattedDuration})
        </span>
      </h3>
      <ul>
        {tracks.map((track, index) => (
          <span>{track}</span>
        ))}
      </ul>
    </section>
  )
};

TrackList.propTypes = {
  name: PropTypes.string.isRequired,
  tracks: PropTypes.array.isRequired,
  formattedDuration: PropTypes.string.isRequired,
  isTooLong: PropTypes.bool.isRequired
}

export default TrackList;
