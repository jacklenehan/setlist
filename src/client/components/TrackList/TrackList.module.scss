.root {
  display: flex;
  flex-flow: column nowrap;
  align-items: flex-start;
  padding: 0 0.5rem;

  ul {
    width: 100%;
    list-style-type: none;
    margin: 0;
    padding: 0;
  }
}

.tooLong {
  color: red;
}

.set {
  grid-column-start: 3;
  grid-column-end: 7;

  &.importFormHidden {
    grid-column-start: 1;
    grid-column-end: 6;
  }
}

.reserve {
  grid-column-start: 7;
  grid-column-end: 11;

  &.importFormHidden {
    grid-column-start: 6;
    grid-column-end: 11;
  }
}
