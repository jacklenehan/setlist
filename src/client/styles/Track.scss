.Track {
  display: flex;
  cursor: move;
  margin: .25em 0;
  padding: .25em;
  justify-content: space-between;

  .track-info {
    display: flex;
  }

  .bpm {
    margin-right: .5em;
    width: 3em;
    background: #ddeaff;
    border: none;
    font-size: 14px;

    &:focus {
      outline: none;
    }
  }

  span {
    display: block;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    flex-grow: 1;
  }

  button {
    width: 2rem;
    padding: 0;
    flex-shrink: 0;
  }

  &:first-child {
    margin-top: 0;
  }

  &:last-child {
    margin-bottom: 0;
  }

  &.dragging {
    opacity: 0.25;
  }

  &.song {
    background: #f0f0f0;
  }

  &.spacer {
    background: #ffffff;

    .track-info {
      text-align: center;
    }

    .bpm {
      opacity: 0;
    }
  }
}
