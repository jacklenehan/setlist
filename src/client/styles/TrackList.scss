.TrackList {
  display: flex;
  flex-flow: column nowrap;
  align-items: flex-start;
  padding: 0 .5rem;

  ul {
    width: 100%;
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .duration.too-long {
    color: red;
  }

  &.set {
    grid-column-start: 3;
    grid-column-end: 7;
  }

  &.reserve {
    grid-column-start: 7;
    grid-column-end: 11;
  }
}
