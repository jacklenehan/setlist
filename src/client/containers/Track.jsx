import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import '../styles/Track.css';

let Track = ({ id }) => (
  <li className={`Track`}>
    <input
      placeholder='bpm'
      className='bpm'
      type='text'
    />
    <span>(0) {id}</span>
    <button onClick={() => {}}>></button>
    <button onClick={() => {}}>x</button>
  </li>
);

Track.propTypes = {
  id: PropTypes.string.isRequired
};

Track = connect()(Track);

export default Track;
