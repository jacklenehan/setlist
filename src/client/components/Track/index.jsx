import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import Moment from 'moment'
import { Draggable } from 'react-beautiful-dnd'

import { trackActions } from 'Client/redux/track'
import { setActions } from 'Client/redux/set'
import { storeTempoLocally } from 'Client/helpers'

import styles from './Track.module.scss'

class Track extends React.Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    removeTrackFromList: PropTypes.func.isRequired,
    addTrackToList: PropTypes.func.isRequired,
    changeTrackTempo: PropTypes.func.isRequired,
    index: PropTypes.number.isRequired,
    data: PropTypes.shape({
      artist: PropTypes.string.isRequired,
      tempo: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      durationMs: PropTypes.number.isRequired,
    }).isRequired,
  }

  handleChange = (e) => {
    const { changeTrackTempo, id } = this.props

    e.preventDefault()
    const {
      target: { value },
    } = e
    const notDigits = /[^0-9]/

    if (value.length > 3) return
    if (notDigits.test(value)) return

    changeTrackTempo(id, value)
    storeTempoLocally(id, value)
  }

  render() {
    const {
      removeTrackFromList,
      index,
      set,
      id,
      data: { artist, tempo, name, durationMs },
    } = this.props

    return (
      <Draggable key={id} draggableId={id} index={index}>
        {(provided) => (
          <li
            className={styles.root}
            ref={provided.innerRef}
            {...provided.draggableProps}
            {...provided.dragHandleProps}
          >
            <input
              placeholder="tempo"
              className={styles.tempo}
              type="text"
              value={tempo}
              onChange={this.handleChange}
            />
            <span>
              ({Moment.duration(durationMs).format('m:ss')}) {artist} - {name}
            </span>
            <button
              type="button"
              onClick={() => {
                removeTrackFromList(id, set)
              }}
            >
              x
            </button>
          </li>
        )}
      </Draggable>
    )
  }
}

function mapStateToProps(state, ownProps) {
  return {
    data: {
      ...state.tracks.tracks[ownProps.id],
    },
  }
}

export default connect(
  mapStateToProps,
  {
    removeTrackFromList: setActions.removeTrackFromList,
    addTrackToList: setActions.addTrackToList,
    changeTrackTempo: trackActions.changeTrackTempo,
  },
)(Track)
