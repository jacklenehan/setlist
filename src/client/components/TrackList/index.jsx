import React from 'react'
import PropTypes from 'prop-types'
import Moment from 'moment'
import { connect } from 'react-redux'
import cn from 'classnames'

import Track from 'Client/components/Track'

import styles from './TrackList.module.scss'

const TrackList = ({ id, maxDuration, totalDuration, tracks, importFormHidden }) => {
  const maxDurationInSeconds = Moment.duration(maxDuration).asMilliseconds()
  const formattedDuration = Moment.duration(totalDuration).format('h:mm:ss', { trim: false })

  return (
    <section className={cn(styles.root, styles[id], { [styles.importFormHidden]: importFormHidden })}>
      <h3>
        {id}{' '}
        <span className={cn({ [styles.tooLong]: totalDuration > maxDurationInSeconds })}>({formattedDuration})</span>
      </h3>
      <ul>
        {tracks.map((track) => (
          <Track key={track} id={track} list={id} />
        ))}
      </ul>
    </section>
  )
}

TrackList.propTypes = {
  id: PropTypes.string.isRequired,
  maxDuration: PropTypes.string.isRequired,
  totalDuration: PropTypes.number.isRequired,
  tracks: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  importFormHidden: PropTypes.bool,
}

TrackList.defaultProps = {
  importFormHidden: false,
}

function mapStateToProps(state, ownProps) {
  return {
    totalDuration: ownProps.tracks.reduce((duration, track) => duration + state.tracks.tracks[track].durationMs, 0),
  }
}

export default connect(mapStateToProps)(TrackList)
