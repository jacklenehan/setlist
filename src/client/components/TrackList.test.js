import React from 'react';
import { shallow } from 'enzyme';
import TrackList from './TrackList';
import Track from '../containers/Track';

describe('TrackList', () => {
  let props;
  let wrapper;

  const trackList = () => {
    if(!wrapper) {
      wrapper = shallow(
        <TrackList {...props} />
      );
    }
    return wrapper;
  }

  beforeEach(() => {
    props = {
      formattedDuration: 'foo',
      listTracks: [],
      isTooLong: false,
      list: 'bar'
    };

    wrapper = undefined;
  });

  describe('render', () => {
    it('renders without crashing', () => {
      trackList();
    });

    it('renders a top-level <section>', () => {
      expect(trackList().is('section')).toBe(true);
    });

    describe('props.duration behavior', () => {
      it('indicates the list duration', () => {
        expect(trackList().find('.duration').text()).toContain(props.formattedDuration);
      });

      describe('when duration is greater than maxDuration', () => {
        beforeEach(() => {
        });

        it('indicates that the duration is too long', () => {
          props.isTooLong = true;
          expect(trackList().find('.duration').is('.too-long')).toBe(true);
        });
      });
    });

    describe('props.list behavior', () => {
      it('includes the list as a selector', () => {
        expect(trackList().is(`.${props.list}`)).toBe(true);
      });

      it('includes the list as an h3', () => {
        expect(trackList().find('h3').text()).toContain(props.list)
      });
    });

    describe('props.tracks behavior', () => {
      it('renders a <Track> for each track', () => {
        props.listTracks = [{id: 1}, {id: 2}];
        expect(trackList().find(Track).length).toBe(2);
      });
    });
  });
});
