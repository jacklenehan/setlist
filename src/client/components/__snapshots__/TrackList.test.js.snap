// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackList render renders correctly 1`] = `
<section
  className="TrackList bar"
>
  <h3>
    bar
     
    <span
      className="qunx"
    >
      (
      foo
      )
    </span>
  </h3>
  <ul />
</section>
`;
